# Author Info
---
Name: **Ejaz-ur-Rehman**\
Business Unit Head | Data Analyst | Crystal Tech (Project of MUZHAB Group)\
Faculty Member | PGP Corordinator | Department of Management Sciences | Al Ghazali University\
MS (Finance), Associate Data Scientist (Python), PhD Scholar\
Karachi, Pakistan

![Date](https://img.shields.io/badge/Date-8--August--2026-green?logo=google-calendar)
[![Email](https://img.shields.io/badge/Email-ijazfinance%40gmail.com-blue?logo=gmail)](mailto:ejaz.rehman@agu.edu.pk)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-Ejaz--ur--Rehman-blue?logo=linkedin)](https://www.linkedin.com/in/ejaz-ur-rehman/)
[![GitHub](https://img.shields.io/badge/GitHub-ejazurrehman-black?logo=github)](https://github.com/ejazurrehman)

# Student Instructions
## Islamic FINTECH Final Practical Examination
### Examination Duration: 3 Hours
#### Total Marks: 40 Marks

# Instructions
1. Read all questions carefully before attempting the examination.
2. This is an open-book practical examination. You may use your course notes, lecture slides, and other course materials.
3. You may use only your own laptop for the examination.
4. Internet access may be used only for Python documentation and package references, The use of AI tools (e.g., ChatGPT, Copilot, Gemini, Claude, etc.) during the examination are allowed.
5. Complete all tasks in the provided Jupyter Notebook (.ipynb) file.
6. Write your Python code only in the designated code cells. Do not modify, delete, or rearrange the existing notebook structure.
7. Wherever an Interpretation or Business Insight section is provided, explain your findings in clear and concise language. Marks will be awarded for both technical accuracy and analytical reasoning.
8. Unless otherwise instructed, all charts must include appropriate titles and axis labels.
9. Ensure that your notebook executes without errors before submission.
10. Save your notebook using the following naming convention:
    
    Student_Name_Student_ID.ipynb

11. Uploading link will provide you druing the exam, you can upload your NB in between 4:00 PM to 5:00 PM

> Important: Execute each code cell after writing your solution. Only executed cells with visible outputs will be considered for evaluation where applicable.


# Exploratory Data Analysis and Feature Engineering of the Islamic Banking Dataset for GCC Countries
## Background

> Islamic banks generate large volumes of financial and regulatory data that support strategic pl nning, risk management, regulatory compliance, and business decision-making. Before such data can be used for advanced analytics or machine learning applications, it must be carefully examined, cleaned, and transformed into a reliable analytical dataset.

> In this examination, you will work with a Global Islamic Banking Dataset, which contains financial and regulatory indicators of Islamic banking institutions across GCC countries. The dataset has been compiled from publicly available Islamic banking statistics and includes variables related to banking performance, assets, financing, deposits, capital adequacy, and other key financial indicators.

> Your task is to perform a complete Exploratory Data Analysis (EDA) and Feature Engineering process to prepare the dataset for future predictive modelling and business intelligence applications.

## Your Role

> Assume that you have recently joined the Data Analytics Department of an Islamic Bank as a Junior Data Analyst. Your manager has assigned you the responsibility of evaluating the quality of the available banking data before it is used for developing Artificial Intelligence and Machine Learning models. You are expected to investigate the dataset, identify potential data quality issues, perform appropriate preprocessing, generate meaningful visualizations, engineer useful features, and communicate your findings through business-oriented interpretations.

## Project Objectives

During this examination, you are expected to:

- Import and configure the required Python libraries.
- Load and inspect the Islamic Banking dataset.
- Assess the overall quality of the data.
- Identify and handle missing values, duplicate records, and inconsistent data.
- Perform Exploratory Data Analysis (EDA) using appropriate statistical summaries and visualizations.
- Engineer meaningful features that could improve future machine learning models.
- Interpret analytical results from both technical and business perspectives.
- Present concise recommendations based on your analysis.

## Data Source

The dataset used in this examination is based on data from the Islamic Financial Services Board (IFSB) and its Prudential and Structural Islamic Financial Indicators (PSIFIs) database.

The IFSB Data Portal provides prudential and structural financial indicators for the Islamic Financial Services Industry. The PSIFIs are designed to provide internationally comparable information on the financial soundness, growth, structure, and risk characteristics of Islamic banking systems.

Official Data Source: https://data.ifsb.org/?utm_source=chatgpt.com

## Data Scope
For this examination, the dataset has been prepared for practical analysis and contains observations relating to:

- GCC countries
- Quarterly reporting periods
- Islamic banking indicators
- Financial values
- USD exchange rates
- Values reported in USD millions

> The dataset has been prepared and provided as part of this examination. Students are not required to download or collect additional data.

## Expected Deliverables

By the end of this examination, your submitted notebook should demonstrate your ability to:

- Understand the structure and characteristics of a real-world financial dataset.
- Apply appropriate data preprocessing techniques.
- Generate meaningful statistical and graphical analyses.
- Develop informative features from existing variables.
- Communicate findings in a professional and business-oriented manner.
- Produce clean, reproducible, and well-documented Python code.

## Note to Students

> This examination assesses not only your ability to write Python code but also your ability to interpret data, think analytically, and apply data science techniques to solve practical problems in the Islamic banking industry.

# Section 0
Import required Libraries

In [2]:
# ============================================================
# Import Required Libraries
# (Pre-configured by the Instructor)
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

plt.style.use("default")

> Note: The required libraries and notebook configuration have been preloaded. Students are not required to modify this section.

In [3]:
# ============================================================
# Load the Dataset (Pre-configured by the Instructor)
# ============================================================

df = pd.read_csv("Islamic_Banking_GCC_Data.csv", encoding="cp1252")

# Display confirmation
print("Dataset loaded successfully.")

Dataset loaded successfully.


## Section 1: Understanding the Dataset (05 Marks)
Objective

The first step in any data analytics project is to understand the structure, composition, and characteristics of the dataset. Before performing data cleaning or analysis, a data analyst must inspect the dataset to identify the available variables, their data types, and the overall quality of the data.

## Student Instructions

Using the pre-loaded dataset `(df)`, perform an initial exploration of the data by completing the following tasks.

### Task 1.1: Display the first five observations of the dataset.

In [ ]:
# Display first five rows


### Task 1.2: Display the last five observations of the dataset.

In [ ]:
# Display last five rows


### Task 1.3: Determine the number of rows and columns in the dataset.

In [ ]:
# Dataset dimensions


### Task 1.4: Determine the columns in the dataset.

In [ ]:
# Column names


### Task 1.5: Examine the data types.

In [ ]:
# Data types


### Task 1.6: Generate descriptive statistics.

In [ ]:
# Descriptive statistics


### Task 1.7: Identify missing values.

In [ ]:
# Missing values


## Answer the following Quesitons

### 1. What is the size of the dataset?

### 2. Which variables appear to be categorical?

### 3. Which variables may require preprocessing?

### 4. Write three observations.

## Section 2: Data Quality Assessment (05 Marks)

### Objective

Before performing further analysis, assess the quality and reliability of the Islamic Banking dataset. Identify potential issues that may affect subsequent analysis and machine learning applications.

In [ ]:
# ============================================================
# Section 2: Data Quality Assessment
# Total Marks: 05
# ============================================================

## Student Instructions

Using the pre-loaded dataset `(df)`, perform the following tasks.

### Task 2.1: Identify the number of missing values in each variable.

In [ ]:
# Identify the number of missing values in each variable


### Task 2.2: Calculate the percentage of missing values for each variable.

In [ ]:
# Calculate the percentage of missing values



### Task 2.3: Identify Duplicate Records: Determine whether the dataset contains duplicate records and report the total number of duplicate observations.

In [ ]:
# Identify duplicate records



### Task 2.4: Examine Data Types: Examine the data type of each variable in the dataset.

In [ ]:
# Examine data types



### Task 2.5: Identify Inappropriate Data Types: Based on your examination of the data types, identify any variable whose current data type appears inappropriate for analysis.

Provide a brief justification for your answer.

In [ ]:
# Display data types together with sample values
# to support identification of potentially inappropriate types


### Task 2.6: Examine Unique and Unusual Values

Identify variables containing unique, unexpected, or unusual values that may require further investigation.

Focus particularly on categorical variables.

In [ ]:
# Examine unique values of categorical variables


### For frequency-based inspection:

In [18]:
for col in categorical_columns:
    print(f"\n--- {col} ---")
    display(df[col].value_counts(dropna=False))


--- Country ---


Country
Oman                    17329
Saudi Arabia            16940
United Arab Emirates    10417
Kuwait                   8228
Bahrain                  8106
Qatar                    6112
Name: count, dtype: int64


--- Regional Code ---


Regional Code
GCC    67132
Name: count, dtype: int64


--- Report Type ---


Report Type
Islamic Banking            49036
Islamic Banking Windows    18096
Name: count, dtype: int64


--- Time Period ---


Time Period
2021Q4    1726
2022Q1    1726
2022Q2    1726
2022Q3    1726
2022Q4    1726
2016Q4    1534
2017Q2    1534
2017Q3    1534
2019Q4    1534
2018Q2    1534
2018Q3    1534
2018Q1    1534
2017Q4    1534
2020Q2    1534
2020Q3    1534
2020Q1    1534
2017Q1    1534
2021Q3    1534
2021Q2    1534
2021Q1    1534
2020Q4    1534
2023Q1    1534
2023Q2    1534
2023Q3    1534
2018Q4    1534
2019Q1    1534
2019Q2    1534
2019Q3    1534
2013Q4    1343
2014Q2    1343
2014Q3    1343
2014Q1    1343
2015Q4    1343
2016Q2    1343
2016Q1    1343
2016Q3    1343
2014Q4    1343
2015Q1    1343
2015Q2    1343
2015Q3    1343
2023Q4    1341
2024Q2     956
2024Q1     956
2024Q3     956
2025Q2     772
2025Q3     772
2024Q4     579
2025Q1     579
2025Q4     193
Name: count, dtype: int64


--- USD Exchange Rate ---


USD Exchange Rate
3.75    16304
0.38    16171
3.67     9264
         7720
3.64     6112
0.3      4675
0.31     2618
0.39     1544
3.67      960
0.28      935
3.75      636
4.67      193
Name: count, dtype: int64


--- Indicator Code ---


Indicator Code
CP01a                350
CP01a_010            350
CP01a_020            350
CP02a                350
CP02a_010            350
CP02a_020            350
CP03a                350
CP03a_010            350
CP03a_020            350
CP01b                350
CP01b_010            350
CP01b_020            350
CP02b                350
CP02b_010            350
CP02b_020            350
CP03b                350
CP03b_010            350
CP03b_020            350
CP04                 350
CP04_010             350
CP04_020             350
CP05                 350
CP05_010             350
CP05_020             350
CP06                 350
CP06_010             350
CP06_020             350
CP07                 350
CP07_010             350
CP07_020             350
CP08                 350
CP08_010             350
CP08_020             350
CP09                 350
CP09_010             350
CP09_020             350
CP10                 350
CP10_020             350
CP11                 350
CP11_010  


--- Indicator Description ---


Indicator Description
Total regulatory capital                                                                                                                      1750
RWA                                                                                                                                           1750
Tier 1 capital                                                                                                                                1400
Total assets                                                                                                                                  1050
Others                                                                                                                                        1006
Net income (before extraordinary items, taxes, and Zakat)-Annualised                                                                           700
Gross income                                                                                    


--- currency ---


currency
NC     48628
       11964
USD     6480
n.a       60
Name: count, dtype: int64


--- units ---


units
M      36013
K      19085
       10680
n.a     1354
Name: count, dtype: int64


--- Values ---


Values
                22871
0                6493
4                  84
6                  81
0.2                76
2                  69
8                  65
1                  62
1.3                52
0.91               52
0.44               49
1.5                49
4.93               47
5.99               47
0.54               46
29.6               46
17.14              46
6.22               46
5.13               45
10.99              45
7.62               45
4.55               45
22.98              44
5.29               44
0.58               44
3.69               44
21                 26
100                26
5                  25
7                  24
10                 22
32                 21
23                 19
16                 17
963                17
9                  16
14                 16
55                 14
64                 14
15                 13
53                 13
20                 13
24                 13
179                13
17                 12
902


--- Values in USD Millions ---


Values in USD Millions
               23092
0               6537
4                 82
6                 72
8                 64
0.27              53
2                 46
0.66              31
0.74              28
0.53              28
1.91              26
0.65              24
100               24
1.44              24
0.05              23
1.79              22
0.09              22
0.21              22
32                21
0.06              21
1.42              21
2.5               21
0.33              20
21                20
5                 19
0.04              19
0.19              19
0.02              18
0.03              18
0.44              17
7                 17
1.87              17
1.77              16
1.9               16
3.01              16
0.1               16
17.46             16
1.02              16
0.12              16
0.8               15
0.08              15
1.33              15
0.37              15
1                 15
1.8               15
4.99              14
1.54       

## Section 3: Data Cleaning (05 Marks)

### Objective

Prepare the dataset for further analysis by correcting formatting issues, handling missing values appropriately, converting variables to suitable data types, and verifying the quality of the cleaned dataset.

### Task 3.1: Standardize Text Variables

Remove unnecessary leading and trailing spaces from the categorical/text variables.

In [19]:
# Identify text variables
text_columns = df.select_dtypes(include="object").columns

# Remove leading and trailing spaces
for col in text_columns:
    df[col] = df[col].str.strip()

### Task 3.2: Convert Blank Values to Missing Values

The dataset may contain blank strings that are not automatically recognized as missing values.Identify blank or empty values and convert them into appropriate missing-value representations.

In [ ]:
# Replace blank or whitespace-only strings with NaN


In [ ]:
# Display the number of missing values in each variable after cleaning


### Task 3.3: Convert Numerical Variables to Appropriate Data Types

Identify the financial variables that should be numerical and convert them to appropriate numeric data types.

The following variables should be examined:

- USD Exchange Rate
- Values
- Values in USD Millions

In [ ]:
# Convert numeric columns to appropriate data types
numeric_columns = [
    "USD Exchange Rate",
    "Values",
    "Values in USD Millions"
]

for col in numeric_columns:
    df[col] = pd.to_numeric(df[col], errors="coerce")

In [ ]:
# Display the data types of the numeric columns after conversion
display(df[numeric_columns].dtypes)

USD Exchange Rate         float64
Values                    float64
Values in USD Millions    float64
dtype: object

### Task 3.4: Handle Duplicate Records

> Check the dataset for duplicate observations.

> If duplicate records are present, remove them appropriately. If no duplicates are present, report the result.

In [ ]:
# Identify and remove duplicate records


# Remove duplicate records


### Task 3.5: Review the missing values remaining after the previous cleaning steps.

For each type of missing value, determine an appropriate treatment based on the nature of the variable.

> Important: Do not automatically replace all missing values with the mean, median, or zero. Your treatment should be justified.

In [ ]:
# Find the missing values summary after cleaning and removing duplicates


In [26]:
# Keep missing observations at this stage where
# missingness may represent unreported/not-applicable data.

print("Missing values retained for appropriate analysis.")

Missing values retained for appropriate analysis.


### Task 3.6: Verify the Cleaned Dataset

After completing the cleaning process, verify that the dataset is ready for further analysis.

Check:

- Data types
- Remaining missing values
- Duplicate records
- Dataset dimensions

In [ ]:
# Show the final dataset shape, data types, missing values, and duplicate records after cleaning


## Section 4: Univariate Analysis (05 Marks)

### Objective

Perform univariate analysis to understand the distribution, frequency, and basic characteristics of individual categorical and numerical variables in the Islamic Banking dataset.

### Task 4.1: Identify Categorical and Numerical Variables

Identify the variables that are:

- Categorical
- Numerical

In [ ]:
# Identify categorical variables


# Identify numerical variables


### Task 4.2: Analyse Categorical Variables

For each categorical variable, determine the frequency of its categories.

Display the 10 most frequent categories for each categorical variable.

In [ ]:
# Display the categorical and numerical variables


### Task 4.3: Analyse Numerical Variables

Generate descriptive statistics for the numerical variables, including:

- Count
- Mean
- Standard deviation
- Minimum
- Quartiles
- Maximum

In [ ]:
# Display the summary statistics for numerical variables


### Task 4.4: Identify Dominant Categories

Identify the most frequently occurring category for the major categorical variables.

For example:

- Country
- Report Type
- Currency
- Units

In [ ]:
# Display the top category and its frequency for each categorical variable


### Task 4.5: Interpret the Numerical Variables

Based on the descriptive statistics, identify:

1. The numerical variable with the highest variability.
2. The variable with the largest range between minimum and maximum values.
3. Any variable that appears to contain potentially extreme observations.

In [ ]:
# Calculate the range (max - min) for each numeric variable


In [ ]:
# Calculate the standard deviation for each numeric variable to assess variability


### Interpretation

## Section 5: Data Visualizaiton (05 Marks)

### Objective

Use appropriate graphical techniques to visualize the distribution and characteristics of variables in the Islamic Banking dataset. Interpret the visualizations from both statistical and business perspectives.

### Task 5.1: Country Distribution

Create a bar chart showing the top 10 countries by number of observations in the dataset.

Your chart must include:

- Appropriate title
- X-axis label
- Y-axis label
- Readable category labels

In [ ]:
# Display the distribution of the top 10 countries by number of observations using a bar plot




### Interpretation

### Task 5.2: Financial Value Distribution

The variable Values in USD Millions contains observations with substantially different magnitudes. Create an appropriate visualization to examine its distribution.

You should:

1. Examine the distribution of the variable.
2. If the distribution is highly skewed, apply an appropriate transformation for visualization.
3. Create a histogram of the transformed values.
4. Clearly label the chart.

In [ ]:
# Create a log-transformed version for visualization



### Interpretation

### Task 5.3: Identify Potential Outliers

Create a boxplot of the log-transformed financial values created in Task 5.2 to investigate potential extreme observations.

In [ ]:
# Create a boxplot for the "Values in USD Millions" variable to visualize its distribution and identify potential outliers


### Interpretation


1. Does the boxplot indicate potential outliers?
2. Why should an observation identified as an outlier not automatically be removed from a financial dataset?

### Task 5.4: Comparative Financial Analysis
Analyse the distribution of Total Financing across GCC countries.

Using the Indicator Description variable:

- Filter the dataset to include only observations where the indicator is Total financing.
- Calculate the average value of Total Financing in USD Millions for each country.
- Create a horizontal bar chart comparing the average Total Financing across the six GCC countries.
- Sort the countries from lowest to highest average Total Financing.
- Add an appropriate title and axis labels.

In [ ]:
# Filter the dataset for Total Financing


# Calculate average financing by country


In [ ]:
# Create a horizontal bar plot for average financing by country


## Section 6: Time Analysis (05 Marks)

### Objective

Analyse the movement of Islamic banking financing over time using the quarterly reporting periods available in the dataset.

### Task 6.1: Examine the Reporting Period

Examine the Time Period variable and identify:

- The earliest reporting period.
- The latest reporting period.
- The total number of reporting periods.



In [ ]:
# Number of unique reporting periods


# Earliest reporting period


# Latest reporting period


### Task 6.2: Prepare the Time Period Variable

The Time Period variable is recorded in quarterly format, such as:
> Convert the Time Period variable into a quarterly pandas Period format and create a new variable named Quarter

In [ ]:
# Create a new column for the quarter of each reporting period and display the first five rows of the updated dataset


### Task 6.3: Analyse Total Financing Over Time

Using Total financing from Indicator Description:

- Filter the dataset for Total Financing.
- Calculate the average Total Financing in USD Millions for each reporting period.
- Display the results in chronological order.

In [ ]:
# Filter the dataset for Total Financing
financing_time = df[
    df["Indicator Description"] == "Total financing"
].copy()

# Convert financial value to numeric
financing_time["Values in USD Millions"] = pd.to_numeric(
    financing_time["Values in USD Millions"],
    errors="coerce"
)

# Average financing by reporting period
quarterly_financing = (
    financing_time
    .groupby("Quarter")["Values in USD Millions"]
    .mean()
    .sort_index()
)

display(quarterly_financing)

Quarter
2013Q4    4.483943e+04
2014Q1    4.698505e+04
2014Q2    4.995076e+04
2014Q3    5.165157e+04
2014Q4    5.287042e+04
2015Q1    5.413258e+04
2015Q2    5.593970e+04
2015Q3    5.644989e+04
2015Q4    5.838034e+04
2016Q1    6.071590e+04
2016Q2    6.246147e+04
2016Q3    6.260143e+04
2016Q4    6.253777e+04
2017Q1    6.375545e+04
2017Q2    6.507385e+04
2017Q3    6.581785e+04
2017Q4    6.536958e+04
2018Q1    6.594156e+04
2018Q2    6.660625e+04
2018Q3    6.757040e+04
2018Q4    6.788843e+04
2019Q1    6.869133e+04
2019Q2    6.978144e+04
2019Q3    7.152770e+04
2019Q4    7.345260e+04
2020Q1    7.581685e+04
2020Q2    7.825034e+04
2020Q3    8.021770e+04
2020Q4    8.257352e+04
2021Q1    8.558578e+04
2021Q2    8.798604e+04
2021Q3    8.984732e+04
2021Q4    8.597199e+04
2022Q1    8.941236e+04
2022Q2    9.198395e+04
2022Q3    9.421967e+04
2022Q4    9.922823e+04
2023Q1    1.078460e+05
2023Q2    1.097905e+05
2023Q3    1.119314e+05
2023Q4    1.178285e+05
2024Q1    1.695065e+05
2024Q2    1.746347e+05
202

### Task 6.4: Visualize the Financing Trend

Create a line chart showing the average Total Financing over the reporting periods.

Your chart must include:

- Appropriate title
- X-axis label
- Y-axis label
- Clearly ordered time periods

In [ ]:
# Create a line plot for average financing over time
plt.figure(figsize=(12, 5))

# Plotting the average total financing over time
plt.plot(
    quarterly_financing.index.astype(str),
    quarterly_financing.values,
    marker="o"
)

plt.title("Average Total Financing Over Time")
plt.xlabel("Reporting Period")
plt.ylabel("Average Total Financing (USD Millions)")

plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

### Task 6.5: Interpret the Trend

Based on your analysis and visualization, answer the following:

1. What overall trend do you observe in Total Financing?
2. Are there periods of noticeable increase or decrease?
3. What does the trend suggest about the development of Islamic banking financing over the observed period?

## Section 7: Correlation Analysis (05 Marks)

### Objective

Examine the relationships between numerical variables in the Islamic Banking dataset using correlation analysis and interpret the strength and direction of those relationships.

### Task 7.1: Prepare Numerical Variables

Identify all numerical variables that can be included in the correlation analysis.

In [ ]:
# Identify numerical variables
num_cols = df.select_dtypes(include=np.number).columns.tolist()

print("Numerical Variables:")
print(num_cols)

### Task 7.2: Generate the Correlation Matrix

Calculate the Pearson correlation matrix for the numerical variables.

Display the resulting correlation matrix.

In [ ]:
# Calculate Pearson correlation matrix
corr = df[num_cols].corr(method="pearson")

display(corr)

### Task 7.3: Visualize the Correlation Matrix

Create a correlation heatmap using the numerical variables.

Your heatmap should:

- Display correlation coefficients.
- Use an appropriate colour scale.
- Include a meaningful title.
- Clearly display variable names.

In [ ]:
# Create a heatmap to visualize the correlation matrix


### Task 7.4: Identify Strong Relationships

Using the correlation matrix, identify:

- The pair of numerical variables with the strongest positive correlation.
- The pair of numerical variables with the strongest negative correlation, if one exists.
- One pair with a relatively weak correlation.

In [ ]:
# Extract the upper triangle of the correlation matrix
upper = corr.where(
    np.triu(np.ones(corr.shape), k=1).astype(bool)
)

# Convert correlations into a long-format table
corr_pairs = (
    upper
    .stack()
    .reset_index()
)

corr_pairs.columns = ["Variable 1", "Variable 2", "Correlation"]

# Sort by correlation
corr_pairs_sorted = corr_pairs.sort_values(
    "Correlation",
    ascending=False
)

display(corr_pairs_sorted.head(10))

### Strongest positive correlation

In [ ]:
strongest_positive = corr_pairs.loc[
    corr_pairs["Correlation"].idxmax()
]

display(strongest_positive)

### Strongest negative correlation

In [ ]:
strongest_negative = corr_pairs.loc[
    corr_pairs["Correlation"].idxmin()
]

display(strongest_negative)

### Weakest relationship

In [ ]:
weakest_relationship = corr_pairs.loc[
    corr_pairs["Correlation"].abs().idxmin()
]

display(weakest_relationship)

### Task 7.5: Interpret the Correlations

For the strongest relationship identified in Task 7.4, explain:

1. The direction of the relationship.
2. The strength of the relationship.
3. What the relationship may indicate from an Islamic banking perspective.

| Correlation | Interpretation                 |
| ----------: | ------------------------------ |
| Close to +1 | Strong positive relationship   |
| Close to -1 | Strong negative relationship   |
|  Close to 0 | Weak or no linear relationship |


## Section 8: Feature Engineering (2.5 Marks)

### Objective

Create meaningful new features from the existing Islamic Banking dataset that can support future statistical analysis, machine learning, and business intelligence applications.

### Task 8.1: Create Year and Quarter Features

The dataset contains quarterly reporting periods in the Time Period variable.

Create two new variables:

- Year
- Quarter_Number
- Quatter_Label

In [ ]:
# Create Year and Quarter features
df["Year"] = df["Quarter"].dt.year
df["Quarter_Number"] = df["Quarter"].dt.quarter

In [60]:
# Create a Quarter_Label feature for better visualization and interpretation
df["Quarter_Label"] = "Q" + df["Quarter_Number"].astype(str)

### Verification

In [ ]:
# Display the first 10 rows of the new features
display(
    df[
        ["Time Period", "Year", "Quarter_Number", "Quarter_Label"]
    ].head(10)
)

### Task 8.2: Create a Log-Transformed Financial Feature

The Values in USD Millions variable contains highly skewed financial observations.

Create a new feature called:

Log_Values_USD_Millions

using an appropriate logarithmic transformation.

In [ ]:
# Create a log-transformed financial feature
df["Log_Values_USD_Millions"] = np.log1p(
    df["Values in USD Millions"]
)

### Verification

In [ ]:
# Display the first 10 rows of the original and log-transformed values
display(
    df[
        ["Values in USD Millions", "Log_Values_USD_Millions"]
    ].head(10)
)

### Interpretation

Explain:

1. Why is a logarithmic transformation useful for highly skewed financial data?
2. Why should the original Values in USD Millions variable be retained?

### Task 8.3: Create a Financing Indicator Feature

Create a binary feature called:

Is_Financing

The feature should take:

- 1 if Indicator Description contains the word "financing"
- 0 otherwise

This feature will allow future analysis to distinguish financing-related indicators from other Islamic banking indicators.

In [ ]:
# Create a binary feature indicating whether the observation is related to financing
df["Is_Financing"] = (
    df["Indicator Description"]
    .str.contains("financing", case=False, na=False)
    .astype(int)
)

### Verification

In [ ]:
# Display the distribution of the Is_Financing feature
display(
    df["Is_Financing"]
    .value_counts()
    .sort_index()
)

### Task 8.4: Assess the Engineered Features

Display the following five variables together:

- Time Period
- Year
- Quarter_Number
- Quarter_Label
- Log_Values_USD_Millions
- Is_Financing

In [ ]:
# Display the first 10 rows of the relevant features
display(
    df[
        [
            "Time Period",
            "Year",
            "Quarter_Number",
            "Log_Values_USD_Millions",
            "Is_Financing"
        ]
    ].head(10)
)

## Section 9: Business Insights and Recommendations (2.5 Marks)

### Objective

Translate the results of your data analysis into meaningful business insights relevant to Islamic banking. Your answers must be based on the analysis performed in Sections 1 to 8.

### Task 9.1: Identify Key Findings

Based on your EDA, visualization, time analysis, correlation analysis, and feature engineering, identify three important findings from the dataset.

For each finding:

1. State the finding.
2. Provide supporting evidence from your analysis.
3. Explain why the finding may be relevant to Islamic banking.

### Task 9.2: Identify a Potential Data or Modelling Risk

Based on your analysis, identify one issue that could create a risk for future Machine Learning or predictive analytics.

Possible areas include:

- Missing values
- Highly skewed variables
- Outliers
- Highly correlated variables
- Unequal country representation
- Data-type problems
- Limited observations

You must identify the issue from your own analysis and explain its potential impact.

### Task 9.3: Recommend Additional Data

Suppose the Islamic Bank wants to develop a future predictive analytics model using this dataset.

> Identify three additional variables that you would recommend collecting.

For each variable, explain why it would improve the analytical or predictive capability of the dataset.

### Task 9.4: Management Recommendations

Assume that you are presenting your findings to the Head of Data Analytics of an Islamic Bank.

> Based on your analysis, provide three practical recommendations.

Your recommendations must be:

- Based on evidence from your analysis.
- Relevant to Islamic banking.
- Action-oriented.

### Task 9.5: Executive Summary

> Prepare a concise executive summary of approximately 150 to 200 words covering:

- The purpose of the analysis.
- The most important findings.
- The major data-quality or modelling concern.
- Your key recommendation.

## Congratulations!
You Have Completed Your Islamic FINTECH Final Practical Examination

Congratulations on completing your Islamic FINTECH Final Practical Examination.

> Through this project, you have applied the practical skills developed throughout the course, including Python programming, data cleaning, Exploratory Data Analysis, visualization, time-series analysis, correlation analysis, and feature engineering using Islamic banking data.

> The purpose of this examination was not only to test your ability to write Python code, but also to assess your ability to understand data, identify meaningful patterns, solve analytical problems, and communicate insights from an Islamic banking perspective.

Before submitting, please make sure that:

- All sections have been completed.
- All code cells have been executed.
- Your interpretations and recommendations are included.
- Your notebook is saved with the required filename.
- The required files have been uploaded successfully to the examination portal.

Congratulations once again on completing the course. I wish you success in your academic and professional journey.

## Final Reminder

- Check your notebook.
- Save your work with your name and your student ID
- Submit before the deadline via given upload link on the Folder.